File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here
def first_then_lower_case(list, predicate):
    for element in list:
        if predicate(element):
            return element.lower()
    return None

# Write your powers generator here
def powers_generator(*, base, limit):
    power = 0
    while base ** power <= limit:
        yield base ** power
        power += 1

# Write your say function here
def say(word=None):
    if not hasattr(say, "words"):
        say.words = []
    
    if word is not None:
        say.words.append(word)
        return say  # Return the function itself to allow chaining
    else:
        #return the collected words
        result = " ".join(say.words)
        say.words = []  # Reset
        return result

# Write your line count function here
def meaningful_line_count(filename):
    # Open the file in read mode
    with open(filename, 'r') as file:
        # Read all lines
        lines = file.readlines()

    valid_lines = 0

    for line in lines:
        # Strip leading/trailing whitespaces
        stripped_line = line.strip()

        # Check if the line is not empty, not made up of whitespaces, and does not start with '#'
        if stripped_line and not stripped_line.startswith('#'):
            valid_lines += 1

    return valid_lines

# Write your Quaternion class here
@dataclass(frozen=True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float

    def __add__(self, other):
        return Quaternion(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)

    def __mul__(self, other):
        return Quaternion(
            self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d,
            self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c,
            self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b,
            self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
        )

    def __eq__(self, other):
        return self.a == other.a and self.b == other.b and self.c == other.c and self.d == other.d

    @property
    def coefficients(self):
        return self.a, self.b, self.c, self.d

    @property
    def conjugate(self):
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    def __str__(self):
        components = []
        if self.a != 0:
            components.append(f"{self.a}")
        if self.b != 0:
            components.append(f"{'+' if self.b > 0 else ''}{self.b if abs(self.b) != 1 else '-' if self.b == -1 else ''}i")
        if self.c != 0:
            components.append(f"{'+' if self.c > 0 else ''}{self.c if abs(self.c) != 1 else '-' if self.c == -1 else ''}j")
        if self.d != 0:
            components.append(f"{'+' if self.d > 0 else ''}{self.d if abs(self.d) != 1 else '-' if self.d == -1 else ''}k")

        result = ''.join(components)

        if result.startswith('+'):
            result = result[1:]
        return result if result else '0'

File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(table, predicate)
  for _, str in ipairs(table) do
    if predicate(str) then
      return str:lower()
    end
  end
  return nil
end

-- Write your powers generator here
function powers_generator(base, limit)
  local co = coroutine.create(function()
    local power = 1
    while power <= limit do
      coroutine.yield(power)
      power = power * base
    end
  end)
  return co
end
  
-- Write your say function here
function say(word)
  local words = {}

  local function chainable(next_word)
    if next_word ~= nil then
      table.insert(words, next_word)
      return chainable
    else
      return table.concat(words, " ")
    end
  end

  return chainable(word)
end

-- Write your line count function here
function meaningful_line_count(filename)
  local file = io.open(filename, "r")
  if not file then error("No such file") end

  local count = 0
  for line in file:lines() do
    local trimmed = line:match("^%s*(.-)%s*$")  -- space for leading and trailing
    if trimmed ~= "" and trimmed:sub(1, 1) ~= "#" then
      count = count + 1
    end
  end

  file:close()
  return count
end

-- Write your Quaternion table here
Quaternion = {}
Quaternion.__index = Quaternion

-- Constructor
function Quaternion.new(a, b, c, d)
  return setmetatable({a = a, b = b, c = c, d = d}, Quaternion)
end

-- Addition
function Quaternion.__add(q1, q2)
  return Quaternion.new(q1.a + q2.a, q1.b + q2.b, q1.c + q2.c, q1.d + q2.d)
end

-- Multiplication
function Quaternion.__mul(q1, q2)
  return Quaternion.new(
    q1.a * q2.a - q1.b * q2.b - q1.c * q2.c - q1.d * q2.d,
    q1.a * q2.b + q1.b * q2.a + q1.c * q2.d - q1.d * q2.c,
    q1.a * q2.c - q1.b * q2.d + q1.c * q2.a + q1.d * q2.b,
    q1.a * q2.d + q1.b * q2.c - q1.c * q2.b + q1.d * q2.a
  )
end

-- Coefficients
function Quaternion:coefficients()
  return {self.a, self.b, self.c, self.d}
end

-- Conjugate
function Quaternion:conjugate()
  return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

-- Equality
function Quaternion.__eq(q1, q2)
  return q1.a == q2.a and q1.b == q2.b and q1.c == q2.c and q1.d == q2.d
end

-- String
function Quaternion.__tostring(q)
  local parts = {}

  -- Handle part (a)
  if q.a ~= 0 or (q.b == 0 and q.c == 0 and q.d == 0) then
    table.insert(parts, tostring(q.a))
  end

  -- Handle the 'i' component (b)
  if q.b ~= 0 then
    local bStr = (q.b == 1 and "i" or (q.b == -1 and "-i" or tostring(q.b) .. "i"))
    table.insert(parts, bStr)
  end

  -- Handle the 'j' component (c)
  if q.c ~= 0 then
    local cStr = (q.c == 1 and "j" or (q.c == -1 and "-j" or tostring(q.c) .. "j"))
    table.insert(parts, cStr)
  end

  -- Handle the 'k' component (d)
  if q.d ~= 0 then
    local dStr = (q.d == 1 and "k" or (q.d == -1 and "-k" or tostring(q.d) .. "k"))
    table.insert(parts, dStr)
  end

  -- Combine and add +
  local result = table.concat(parts, "+"):gsub("%+%-", "-")

  -- Remove leading '+' if it's there
  result = result:gsub("^%+", "")

  return result
end


